modules/deployment-model-impl/src/test/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstanceTestCase.java

```diff
@@ -6,70 +6,145 @@
+import java.util.List;
 
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
 
 import org.mockito.ArgumentCaptor;
 
+import io.qameta.allure.Feature;
 import io.qameta.allure.Issue;
+import io.qameta.allure.Story;
 
+@Feature(POLICY_DEPLOYMENT)
+@Story(POLICY_ISOLATION)
+@RunWith(Parameterized.class)
 public class DefaultApplicationPolicyInstanceTestCase extends AbstractMuleTestCase {
 
+  @Parameters(name = "enablePolicyIsolation: {0}")
+  public static Boolean[] parameters() {
+    return new Boolean[] {false, true};
+  }
+
+  public boolean enablePolicyIsolation;
+
+  @Rule
+  public SystemProperty enablePolicyIsolationRule;
+
   private final ServiceRepository serviceRepository = mock(ServiceRepository.class);
 
-  @Test
-  @Issue("MULE-14289")
-  public void correctArtifactTypeForPolicies() throws Exception {
-    MuleContextListener muleContextListener = mock(MuleContextListener.class);
-    ArgumentCaptor<MuleContext> muleContextCaptor = ArgumentCaptor.forClass(MuleContext.class);
+  private PolicyTemplate policyTemplate;
+  private Application application;
+  private Registry appRegistry;
+
+  private MuleContextWithRegistry policyMuleContext;
+  private MuleRegistry policyRegistry;
 
-    PolicyTemplate policyTemplate = mock(PolicyTemplate.class, RETURNS_DEEP_STUBS);
+  public DefaultApplicationPolicyInstanceTestCase(boolean enablePolicyIsolationPropertyValue) {
+    this.enablePolicyIsolation = enablePolicyIsolationPropertyValue;
+    this.enablePolicyIsolationRule =
+        new SystemProperty(ENABLE_POLICY_ISOLATION_PROPERTY, Boolean.toString(enablePolicyIsolationPropertyValue));
+  }
+
+  @Before
+  public void setUp() throws RegistrationException {
+    policyTemplate = mock(PolicyTemplate.class, RETURNS_DEEP_STUBS);
     when(policyTemplate.getArtifactClassLoader().getClassLoader()).thenReturn(this.getClass().getClassLoader());
     when(policyTemplate.getDescriptor()).thenReturn(new PolicyTemplateDescriptor("policyId"));
 
-    Application application = mock(Application.class, RETURNS_DEEP_STUBS);
+    application = mock(Application.class, RETURNS_DEEP_STUBS);
+
     ArtifactContext appArtifactContext = application.getArtifactContext();
     MuleContextWithRegistry appMuleContext = mockContextWithServices();
     when(appArtifactContext.getMuleContext()).thenReturn(appMuleContext);
 
-    Registry registry = appArtifactContext.getRegistry();
-    doReturn(of(appMuleContext)).when(registry).lookupByType(MuleContext.class);
-    doReturn(of(mock(ExtensionManager.class))).when(registry).lookupByName(OBJECT_EXTENSION_MANAGER);
+    appRegistry = appArtifactContext.getRegistry();
+    doReturn(of(appMuleContext)).when(appRegistry).lookupByType(MuleContext.class);
+
+    doReturn(appRegistry).when(application).getRegistry();
+
+    policyRegistry = mock(MuleRegistry.class);
```

Comment:
This is only used here and it is defined as a field. Use it as a local variable.